<h1><center>Support Vector Machines</center></h1>

The goal of this tutorial is to develop your understanding of a basic Machine Learning classification model and learn some basic area of machine learning like supervised learning. It is necessary to not only understand the mathematics behind the models but also to know how to convert that knowledge into code that can work.

### By the end of this tutorial, you’ll have learned:

   - What SVM is
   - How SVM works
   - How to implement KNN in Python, step by step

### Supervised Learning

Supervised learning in machine learning is the task of learning a funcion that maps some inputs to outputs based on some input-output pair. It infers a function from labeled training data consisting of a set of training examples. In supervised learning we are given labeled data with which we can compare our model's predictions with. A supervised learning algorithm analyzes the training data and produces an inferred function, which can be used for mapping new examples.

Classification is an example of supervised learning. We will be looking at one of the most basic classification algorithm called K Nearest Neighbor (KNN) in this tutorial.

#### What is Classification?

In machine learning, classification is the problem of identifying to which of a set of categories a new observation belongs to, on the basis of a training set of data containing observations whose category membership is known (is labeled). For example assigning an email spam or not spam is an example of classification. or Identifying a patient to have cancer or not is also an example of classification.

### Support vector machines

SVM is considered generally to be a classification technique, it but can be used in both types of classification and regression tasks. SVM creates a hyperplane in multi-dimensional space to separate different classes. The main idea of support vector machines is margin maximization that divides data into two classes.

![title](svm.png)

#### Support vectors

These are the points which are closest to the margin. These points will make the decision boundary better by calculating the margins. These support vaectors are more important to the construction of the classifier than other points.

#### Hyperplane

A decision plane which separates the classes.

#### Margin

A gap between the two lines on the closest class points. This is calculated as the perpendicular distance from the line to support vectors or closest points.

#### SVM Kernels

Generally, svm works only for linearly separable classes. To make it work on non-linear data we use something called kernels.
A kernel transforms an input data space into the required space in which data can be separated linearly. SVM uses a technique called the kernel trick. The kernel takes a low-dimensional input space and transforms it into a higher dimensional space. There are different types of kernel like linear kernel, polynomial and rbf kernels.

### Advantages and Disadvantages of KNN

#### Advantages
   - Uses less memory.
   - Works faster for predictions.
   - Works well with a clear margin of separation and with high dimensional space.
   - Offers good accuracy.
    
#### Disadvantages
   - Training time is quite high.
   - Choice of kernels.
   - Outliers can be a problem. 
   - Doesn't work best with overlapping classes.

### Working of SVM

The main objective of SVM is to divide the given data in the best possible way into different classes. The distance between the  nearest data points of each class is known as the margin. The goal of svm is to select a hyperplane with the maximum possible margin between support vectors in the given data. It does it in following ways:
    
  - Creates multiple decision boudaries that divides data into classes in the best way.
  
  ![title](svm2.jpg)
  
  - Now, there ae multiple decision boundaaries but which to select? Intuitively, if we select a hyperplane which is close to the data points of one class, then it might not generalize well. So the goal is to choose the hyperplane which is as far as possible from the data points of each category.
  
  ![title](svm3.jpg)

  - By maximizing the distance between the classes and the hyperplane would result in an optimal separating hyperplane. The goal of SVMs is to find the optimal hyperplane because it not only classifies the existing dataset but also helps predict the class of the unseen data.
  
  ![title](svm.png)

### Implementation using python

These are some of the essential libraries that are needed to use some built-in functionalities provided by python. Some well-known libraries are like sklearn.

In [1]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.metrics import accuracy_score

##### Dataset

For this tutorial we will be working with cancer dataset provided by sklearn library

In [2]:
#Load dataset
cancer = datasets.load_breast_cancer()

Exploring the dataset

In [3]:
# print the names of the 13 features
print("Features: ", cancer.feature_names)

# print the label type of cancer('malignant' 'benign') i.e. 2 classes
print("\nLabels: ", cancer.target_names)

Features:  ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']

Labels:  ['malignant' 'benign']


Let's take a look on targets of data. 

In [4]:
# print the first 50 labels (0:malignant, 1:benign)
print(cancer.target[:50])

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1]


let's split data into training and testing

In [5]:
# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, test_size=0.25)

##### Creating model

Let's start building our svm model by using sklearn.

In [6]:
#Create a svm Classifier
clf = svm.SVC(kernel='linear') # Linear Kernel

#Train the model using the training sets
clf.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

Prediction on test data

In [7]:
#Predict the response for test dataset
y_pred = clf.predict(X_test)

##### Performace

Let's measure performance of our model by comparing predictions with the original labels.

In [8]:
# Model Accuracy
print("Accuracy:", round(accuracy_score(y_test, y_pred)*100),'%')

Accuracy: 94.0 %


Our model got an accuracy of 94 % which is pretty good.